core/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java

```diff
@@ -124,23 +128,26 @@ private void onShardFailure(final int shardIndex, @Nullable ShardRouting shard,
 
     @Override
     public final void run() throws IOException {
-        int shardIndex = -1;
-        for (final SearchShardIterator shardIt : shardsIts) {
-            shardIndex++;
-            final ShardRouting shard = shardIt.nextOrNull();
-            if (shard != null) {
-                performPhaseOnShard(shardIndex, shardIt, shard);
-            } else {
-                // really, no shards active in this group
-                onShardFailure(shardIndex, null, null, shardIt, new NoShardAvailableActionException(shardIt.shardId()));
-            }
+        boolean success = shardExecutionIndex.compareAndSet(0, concurrentRunnables);
+        assert success;
+        for (int i = 0; i < concurrentRunnables; i++) {
+            int index = i;
```

Comment:
let's just use `i` directly? I presume a previous iteration of this work used a lambda that needed a final variable?